c_reference/include/quantized_utils.h

```diff
@@ -53,15 +53,261 @@ void v_q_tanh(const INT_T* const vec, ITER_T len, INT_T* const ret,
+ * @param[out]      ret       pointer to the output matrix
+ * @param[in]       scmat     scaling factor for the input matrix
+ * @param[in]       scvec     scaling factor for the bias vector
+ * @param[in]       scret     scaling factor for the output matrix
+ * @return          none
+ * @example         mat       = {1324, 5453, 3454, 3435, 8789, 3411, 5412, 8934}
+ *                  vec       = {8452, 2341, 9383, 2353}
+ *                  nrows     = 4
+ *                  ncols     = 2
+ *                  ret       = {2775, 3311, 4072, 2305, 6507, 2290, 5051, 5055}
+ *                  scmat     = 1
+ *                  scvec     = 2
+ *                  scret     = 2
+ */
+void m_q_add_vec(const INT_T* const mat, const INT_T* const vec,
+                 ITER_T nrows, ITER_T ncols, INT_T* const ret,
+                 SCALE_T scmat, SCALE_T scvec, SCALE_T scret);
+/**
+ * @brief Performs the column-wise subtraction of a bias term from the input matrix.
+ * dim(mat) = dim(ret) = [nrows][ncols]; dim(vec) = [ncols].
+ * @param[in]       mat       pointer to the input matrix from which subtraction is to be performed
+ * @param[in]       vec       pointer to the bias vector which is to be subtracted
+ * @param[in]       nrows     number of rows of the input matrix
+ * @param[in]       ncols     number of columns of the input matrix
+ * @param[out]      ret       pointer to the output matrix
+ * @param[in]       scmat     scaling factor for the input matrix
+ * @param[in]       scvec     scaling factor for the bias vector
+ * @param[in]       scret     scaling factor for the output matrix
+ * @return          none
+ * @example         mat       = {1324, 5453, 3454, 3435, 8789, 3411, 5412, 8934}
+ *                  vec       = {8452, 2341, 9383, 2353}
+ *                  nrows     = 4
+ *                  ncols     = 2
+ *                  ret       = {-1451, 2141, -618, 1129, 2281, 1120, 361, 3879}
+ *                  scmat     = 1
+ *                  scvec     = 2
+ *                  scret     = 2
+ */
+void m_q_sub_vec(const INT_T* const mat, const INT_T* const vec,
+                 ITER_T nrows, ITER_T ncols, INT_T* const ret,
+                 SCALE_T scmat, SCALE_T scvec, SCALE_T scret);
 // Function for multiplying a matrix with a vector.
 void m_q_mulvec(const INT_T* const mat, const INT_T* const vec, ITER_T nrows,
                 ITER_T ncols, INT_T* const ret, SCALE_T scmat, SCALE_T scvec,
                 SCALE_T H1, SCALE_T H2);
+/**
+ * @brief Performs sparse matrix multiplication of a matrix and a vector.
+ * mat_indices and mat_values combined are a sparse representation; dim(mat_values) = [ndims], dim(mat_indices) = [ndims + ncols].
+ * mat_values[i] is the i^th non-zero value of the input matrix, and mat_indices[i] encodes the location of mat_values[i].
+ * Number of zeroes before Aidx[i] : row of Aval[i]
+ * Aidx[i] + ... + Aidx[l] where l is the largest value less than i such that A[idx] = 0 : column of Aval[i]
```

Comment:
not sure I understand the layout. Is this CSR? where are matrix dimensions in this interface?